### PR TITLE
add remove item from list

### PR DESCRIPTION
**Pull request by changev**
Add two base function 
```removeFromListById```
```removeFromListByIdentifier```

This two functions invoke ```findAndModify``` to atomically run native mongo array operation ```$pullAll```.

### CODE REVIEW

**Comment by changev** [comment]
test this please

**Comment by changev** [comment]
test this please

File: lib/common/model.js [hunk 1]
```diff
@@ -430,6 +438,21 @@ function ModelFactory (
...
         /**
+         * Using an identifier, find document and delete one or more items 
+         * from a list of this document, the values provided specifies the list and items. 
+         * Returns a waterline (bluebird) promise with the document find and modify
+         *
+         * @param {string} identifier
+         * @param {Object} values {index_path: [value]}
+         * @returns {Deferred}
+         */
+        removeFromListByIdentifier: function (identifier, values) {
```
Comment (anhou) [inline lib/common/model.js]: removeItemsFromListByIdentifier or removeListItemsByIdentifier  is better?

File: lib/common/model.js [hunk 2]
```diff
@@ -515,6 +538,25 @@ function ModelFactory (
...
         /**
+         * Using an id, find document and delete one or more items 
+         * from a list of this document, the values provided specifies the list and items. 
+         * Returns a waterline (bluebird) promise with the document find and modify
+         *
+         * @param {string} id
+         * @param {Object} values {index_path: [value]}
+         * @returns {Deferred}
+         */
+        removeFromListById: function (id, values) {
```
Comment (anhou) [inline lib/common/model.js]: removeItemsFromListById or removeListItemsById is better?

File: lib/common/model.js [hunk 3]
```diff
@@ -430,6 +438,21 @@ function ModelFactory (
             });
         },
         /**
+         * Using an identifier, find document and delete one or more items 
+         * from a list of this document, the values provided specifies the list and items. 
+         * Returns a waterline (bluebird) promise with the document find and modify
+         *
+         * @param {string} identifier
+         * @param {Object} values {index_path: [value]}
+         * @returns {Deferred}
```
Comment (anhou) [inline lib/common/model.js]: returns {Prosmise}

File: lib/common/model.js [hunk 4]
```diff
@@ -515,6 +538,25 @@ function ModelFactory (
         },
 
         /**
+         * Using an id, find document and delete one or more items 
+         * from a list of this document, the values provided specifies the list and items. 
+         * Returns a waterline (bluebird) promise with the document find and modify
+         *
+         * @param {string} id
+         * @param {Object} values {index_path: [value]}
+         * @returns {Deferred}
```
Comment (anhou) [inline lib/common/model.js]: returns {Prosmise}

File: lib/common/model.js [hunk 5]
```diff
@@ -260,6 +260,14 @@ function ModelFactory (
                 return self.runNativeMongo('findAndModify', _arguments);
             }
 
+            if (update.$pullAll) {
+                if (!update.$set) {
+                    update.$set = {};
+                }
+                this.addWaterlineCreatedAtToObject(update, '$set');
```
Comment (iceiilin) [inline lib/common/model.js]: might be addWaterlineUpdatedAtToObject()?

File: lib/common/model.js [hunk 6]
```diff
@@ -260,6 +260,14 @@ function ModelFactory (
...
+                return self.runNativeMongo('findAndModify', _arguments);
+            }
+
```
Comment (iceiilin) [inline lib/common/model.js]: The above three code blocks look quite similar. Is it possible to combine them to a simpler one?

File: lib/common/model.js [hunk 7]
```diff
@@ -430,6 +438,21 @@ function ModelFactory (
             });
         },
         /**
+         * Using an identifier, find document and delete one or more items 
+         * from a list of this document, the values provided specifies the list and items. 
+         * Returns a waterline (bluebird) promise with the document find and modify
+         *
+         * @param {string} identifier
+         * @param {Object} values {index_path: [value]}
```
Comment (iceiilin) [inline lib/common/model.js]: it might be better to use the convention in mongoDB as {field: [value]}

File: lib/common/model.js [hunk 8]
```diff
@@ -515,6 +538,25 @@ function ModelFactory (
         },
 
         /**
+         * Using an id, find document and delete one or more items 
+         * from a list of this document, the values provided specifies the list and items. 
+         * Returns a waterline (bluebird) promise with the document find and modify
+         *
+         * @param {string} id
+         * @param {Object} values {index_path: [value]}
```
Comment (iceiilin) [inline lib/common/model.js]: the same as above

File: spec/lib/common/model-spec.js [hunk 9]
```diff
@@ -303,6 +303,78 @@ describe('Model', function () {
...
+            });
+
+            it('should have an updated list', function () {
+                expect(removed.list).to.deep.equal([]);
+            });
+        });
+
+    });
+
+
```
Comment (iceiilin) [inline spec/lib/common/model-spec.js]: Is it necessary to check when the input value is empty?

File: lib/common/model.js (hunk 6 above)
Comment (changev) [inline lib/common/model.js]: Seem that three blocks are easier to read? 
I prefer to maintain them.

**Comment by pengz1** [comment]
+1

**Pull request merged**